rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala

```diff
@@ -2,185 +2,186 @@ package coop.rchain.rspace.test
-    val joinKey = hashChannels(Seq(c))
-    val csKey   = hashChannels(cs)
-    if (_waitingContinuations.get(csKey).forall(_.isEmpty)) {
-      val existing: Seq[Seq[C]] = _joinMap.remove(c).getOrElse(Seq.empty)
-      val filtered              = existing.filter(!_.equals(cs))
-      if (filtered.nonEmpty)
-        _joinMap.put(c, filtered)
+  private[rspace] override def removeJoin(txn: T, c: C, cs: Seq[C]): Unit = {
+    withGNAT(txn, hashChannels(cs)) { (_, csKey, gnatOpt) =>
+      if (gnatOpt.isEmpty || gnatOpt.get.wks.isEmpty) {
+        val existing: Seq[Seq[C]] = _dbJoins.remove(c).getOrElse(Seq.empty)
+        val filtered              = existing.filter(!_.equals(cs))
+        if (filtered.nonEmpty) _dbJoins.put(c, filtered)
+      }
+      gnatOpt
     }
-    collectGarbage(joinKey)
+    collectGarbage(txn, hashChannels(Seq(c)))
   }
 
-  private[rspace] def removeAllJoins(txn: T, c: C): Unit = {
-    _joinMap.remove(c)
-    collectGarbage(hashChannels(Seq(c)))
+  private[rspace] override def removeAllJoins(txn: T, c: C): Unit = {
+    _dbJoins.remove(c)
+    collectGarbage(txn, hashChannels(Seq(c)))
   }
 
-  def close(): Unit = ()
-
-  def getPatterns(txn: T, channels: Seq[C]): Seq[Seq[P]] =
-    _waitingContinuations.getOrElse(hashChannels(channels), Nil).map(_.patterns)
+  override def close(): Unit = ()
 
-  def clear(): Unit = {
-    _keys.clear()
-    _waitingContinuations.clear()
-    _data.clear()
-    _joinMap.clear()
+  override def clear(): Unit = withTxn(createTxnWrite()) { txn =>
+    _dbGNATs.clear()
+    _dbJoins.clear()
     eventsCounter.reset()
   }
 
-  def getStoreCounters: StoreCounters =
-    eventsCounter.createCounters(
-      0,
-      (_keys.size + _waitingContinuations.size + _data.size + _joinMap.size).toLong)
+  override def getStoreCounters: StoreCounters = withTxn(createTxnRead()) { txn =>
+    val gnatsSize = _dbGNATs.foldLeft(0) {
+      case (acc, (_, GNAT(chs, data, wks))) => acc + (chs.size + data.size + wks.size)
+    }
+    eventsCounter.createCounters(0, (gnatsSize + _dbJoins.size).toLong)
+  }
 
-  def isEmpty: Boolean =
-    _waitingContinuations.isEmpty && _data.isEmpty && _keys.isEmpty && _joinMap.isEmpty
+  override def isEmpty: Boolean = withTxn(createTxnRead()) { txn =>
+    _dbGNATs.isEmpty && _dbJoins.isEmpty
+  }
 
-  def toMap: Map[Seq[C], Row[P, A, K]] =
-    _keys.map {
-      case (hash, cs) =>
-        val data = _data.getOrElse(hash, Seq.empty[Datum[A]])
-        val wks  = _waitingContinuations.getOrElse(hash, Seq.empty[WaitingContinuation[P, K]])
+  override def toMap: Map[Seq[C], Row[P, A, K]] = withTxn(createTxnRead()) { txn =>
+    _dbGNATs.map {
+      case (_, GNAT(cs, data, wks)) =>
         (cs, Row(data, wks))
     }.toMap
+  }
+
+  private[this] def isOrphaned(gnat: GNAT[C, P, A, K]): Boolean =
+    gnat.data.isEmpty && gnat.wks.isEmpty && !(gnat.channels.size == 1 && _dbJoins
+      .contains(gnat.channels.head))
```

Comment:
I have to admit that the latter part of this condition makes my head spin a bit.  I realize that it was part of the earlier "garbage collector", so maybe @ys-pyrofex can comment on why this would be necessary.

Specifically, I understand how `data` or `wks` being empty motivates us to flag something as orphaned.  I'm unclear about how the the length of the channels and contents of the join table affects this.  I removed that stuff and tests still pass as well...

Reply:
my goal was to keep the existing implementation as close to the original as possible.

let me see if I can make this go away.